### PR TITLE
Add Mermaid diagram support

### PR DESCRIPTION
**Pull request by WimLeflere**
Add support for generating diagrams in the Mermaid format, which is supported by GitHub.
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams

Example:
```console
$ vcpkg depend-info ableton --mermaid

flowchart TD; ableton --> ableton-link; ableton-link --> asio; ableton-link --> vcpkg-cmake; ableton-link --> vcpkg-cmake-config; asio --> vcpkg-cmake; asio --> vcpkg-cmake-config;
```

When included in Github (inside \`\`\`mermaid ... \`\`\``) it looks like this:
```mermaid
flowchart TD; ableton --> ableton-link; ableton-link --> asio; ableton-link --> vcpkg-cmake; ableton-link --> vcpkg-cmake-config; asio --> vcpkg-cmake; asio --> vcpkg-cmake-config;
```

### CODE REVIEW

**Comment by WimLeflere** [comment]
I'll update #[vcpkg-docs](https://github.com/microsoft/vcpkg-docs) once this is accepted.

File: src/vcpkg/commands.dependinfo.cpp [hunk 1]
```diff
@@ -210,6 +212,22 @@ namespace vcpkg::Commands::DependInfo
...
+        std::string create_mermaid_as_string(const std::vector<PackageDependInfo>& depend_info)
+        {
+            std::string s;
+            s.append("flowchart TD;");
+
+            for (const auto& package : depend_info)
+            {
+                for (const auto& dependency : package.dependencies)
+                {
+                    s.append(Strings::format(" %s --> %s;", package.package, dependency));
```
Comment (BillyONeal) [inline src/vcpkg/commands.dependinfo.cpp:224]:
```suggestion
                    s.append(fmt::format(" {} --> {};", package.package, dependency));
```

File: src/vcpkg/commands.dependinfo.cpp [hunk 2]
```diff
@@ -84,15 +84,17 @@ namespace vcpkg::Commands::DependInfo
...
         constexpr StringLiteral OPTION_MAX_RECURSE = "max-recurse";
         constexpr StringLiteral OPTION_SORT = "sort";
 
         constexpr int NO_RECURSE_LIMIT_VALUE = -1;
 
-        constexpr std::array<CommandSwitch, 3> DEPEND_SWITCHES = {
+        constexpr std::array<CommandSwitch, 4> DEPEND_SWITCHES = {
             {{OPTION_DOT, []() { return msg::format(msgCmdDependInfoOptDot); }},
              {OPTION_DGML, []() { return msg::format(msgCmdDependInfoOptDGML); }},
+             {OPTION_MERMAID, []() { return msg::format(msgCmdDependInfoOptMermaid); }},
```
Comment (BillyONeal) [inline src/vcpkg/commands.dependinfo.cpp:97]: I think you need to add validation that only one of these gets used?

Alternately, add `--format=<dot, or dgml, or mermaid>` and deprecate the old ones to make it clear that only one can be selected?

File: include/vcpkg/base/messages.h [hunk 3]
```diff
@@ -875,6 +875,7 @@ namespace vcpkg
     DECLARE_MESSAGE(CmdDependInfoOptDepth, (), "", "Show recursion depth in output");
     DECLARE_MESSAGE(CmdDependInfoOptDGML, (), "", "Creates graph on basis of dgml");
     DECLARE_MESSAGE(CmdDependInfoOptDot, (), "", "Creates graph on basis of dot");
+    DECLARE_MESSAGE(CmdDependInfoOptMermaid, (), "", "Creates graph on basis of mermaid");
```
Comment (BillyONeal) [inline include/vcpkg/base/messages.h:878]:
```suggestion
    DECLARE_MESSAGE(CmdDependInfoOptMermaid, (), "'mermaid' is a file format type that should remain unlocalized, see https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams", "Creates graph on basis of mermaid");
```

**Review by BillyONeal — CHANGES_REQUESTED** [review]
I think I'm happy with this feature direction.

I would like to see some tests added. (Yes it is bad that DGML and dot don't already have them. But if you want to get out of a hole first you need to put down the shovel...)

File: include/vcpkg/base/messages.h (hunk 3 above)
Comment (BillyONeal) [reply include/vcpkg/base/messages.h:878]: Also should we change "Creates graph on basis of BLAH" to "Create graph in BLAH format" ?

**Comment by WimLeflere** [comment]
> I would like to see some tests added. (Yes it is bad that DGML and dot don't already have them. But if you want to get out of a hole first you need to put down the shovel...)

Any example test that is testing similar functionality?
Should I expose `create_graph_as_string` in `commands.dependinfo.h`, to be able to test it?

**Comment by WimLeflere** [comment]
Where can I find the `format.diff` in the artifacts?

**Comment by BillyONeal** [comment]
> Where can I find the `format.diff` in the artifacts?

Click "Details" on the check run:
Details on the check

Then "View more details on Azure Pipelines":
View more details on Azure Pipelines

Then "1 artifact produced":
1 artifact produced

Then they are there, including `format.diff`:
format.diff

**Comment by BillyONeal** [comment]
@AugP @ras0219-msft @dan-shaw @JavierMatosD and I discussed this yesterday. It was pointed out that there's some overlap with the `--sort` setting because those are kinda like formats too. I'm on the hook to write a one-pager and get our PM friends to OK it.

File: src/vcpkg/commands.dependinfo.cpp [hunk 4]
```diff
@@ -210,17 +244,37 @@ namespace vcpkg::Commands::DependInfo
...
+            }
+
+            return s;
+        }
+
+        std::string create_graph_as_string(OutputFormat output_format,
                                            const std::vector<PackageDependInfo>& depend_info)
         {
-            if (Util::Sets::contains(switches, OPTION_DOT))
+            if (output_format == Dot)
```
Comment (soroshsabz) [inline src/vcpkg/commands.dependinfo.cpp:266]: I suggest use `switch/case` instead of some `if statements`

**Review by soroshsabz — COMMENTED** [review]
I think this PR, needs some end to end tests.

File: src/vcpkg/commands.dependinfo.cpp [hunk 5]
```diff
@@ -210,17 +244,37 @@ namespace vcpkg::Commands::DependInfo
             return s;
         }
 
-        std::string create_graph_as_string(const std::set<std::string, std::less<>>& switches,
+        std::string create_mermaid_as_string(const std::vector<PackageDependInfo>& depend_info)
+        {
+            std::string s;
```
Comment (soroshsabz) [inline src/vcpkg/commands.dependinfo.cpp:249]: I think it is good to add some comments, to describe the flow chart diagram format uses in this method

**Comment by BillyONeal** [comment]
Related: https://github.com/microsoft/vcpkg/issues/30180

Sorry I haven't gotten this yet, hope to look next week.

**Comment by WimLeflere** [comment]
@BillyONeal any progress on the go/nogo for merging the format option with the order option?

**Comment by BillyONeal** [comment]
[Homogenizing dependinfo Switches.pdf](https://github.com/microsoft/vcpkg-tool/files/11520268/Homogenizing.dependinfo.Switches.pdf)

I'm bringing this proposal next week

**Comment by BillyONeal** [comment]
(And yes, I'm signing up to fix the status quo to conform with that so your 'add mermaid' will be a clean add of just that feature rather than asking you to fix our mistakes)

Sorry it takes forever 🥲😭

**Comment by BillyONeal** [comment]
Rest of the maintainer team shot me down on that proposal, and I fixed it to be consistent with their feedback. Updated proposal and implemented in the following PR.

[proposal.pdf](https://github.com/microsoft/vcpkg-tool/files/11551833/proposal.pdf)

Depends on https://github.com/microsoft/vcpkg-tool/pull/1080

File: src/vcpkg/commands.dependinfo.cpp (hunk 5 above)
Comment (BillyONeal) [reply src/vcpkg/commands.dependinfo.cpp]: I think the generated flow chart is explained perfectly well in the docs submission.

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
> The merge looks good to me!

Thanks!

**Pull request merged**